warehouse/query-core/src/main/java/datawave/query/tables/RangeStreamScanner.java

```diff
@@ -135,28 +153,212 @@ public RangeStreamScanner setScannerFactory(ScannerFactory factory) {
          * foo:20130101_0. If our next search space is foo:20130101_0\x00 we will hit all data types within that range...again..and again...and again. To
          * account for this, we put \uffff after the null byte so that we start key is technically the last value within the provided shard, moving us to the
          * exact next key within our RangeStream
          */
-        return new Range(new Key(lastKey.getRow(), lastKey.getColumnFamily(), new Text(lastKey.getColumnQualifier() + "\uffff")), true,
-                        previousRange.getEndKey(), previousRange.isEndKeyInclusive());
+        Key startKey = new Key(lastKey.getRow(), lastKey.getColumnFamily(), new Text(lastKey.getColumnQualifier() + "\uffff"));
+        return new Range(startKey, true, previousRange.getEndKey(), previousRange.isEndKeyInclusive());
+    }
+    
+    /**
+     * Seek range is built with a start key column qualifier set to the seek shard.
+     *
+     * In a worst case scenario the built seek range is for the very next shard, no differently than {@link #buildNextRange}.
+     *
+     * Best case the seek range lies beyond the data available for this scanner, thus negating any need to continue searching within this range stream.
+     *
+     * @param seekShard
+     *            the shard to seek to
+     * @param range
+     *            the RangeStreamScanner's current range
+     * @return the current range with a modified start key
+     */
+    public Range buildSeekRange(String seekShard, Range range) {
+        
+        // Adjust startKey columnQualifier to be the shard.
+        Key startKey = range.getStartKey();
+        Text row = startKey.getRow();
+        Text cf = startKey.getColumnFamily();
+        startKey = new Key(row, cf, new Text(seekShard));
+        
+        return new Range(startKey, true, range.getEndKey(), false);
+    }
+    
+    /**
+     * Seek the underlying iterator to the specified shard.
+     *
+     * First ensures the seek shard does not fall within the currentEntry, resultQueue, or currentQueue.
+     *
+     * The queues are setup not unlike [resultQ.head] - [currentQ.first - currentQ.last] - lastSeenKey
+     *
+     * @param seekShard
+     *            the shard to seek to.
+     * @return the shard we seek'd to.
+     */
+    public String seek(String seekShard) {
+        if (currentEntry == null && resultQueue.isEmpty() && finished) {
+            return null;
+        }
+        
+        String seekedShard = advanceQueues(seekShard);
+        if (seekedShard == null) {
+            
+            this.seekShard = seekShard;
+            this.seeking = true;
+            
+            // Clear queues before calling findTop().
+            this.currentEntry = null;
+            this.resultQueue.clear();
+            this.currentQueue.clear();
+            
+            // Call to hasNext() with empty queues and a null currentEntry triggers a new run of the iterator.
+            if (hasNext()) {
```

Comment:
Although true, this is not intuitive.

Reply:
Per discussion, I'll add a comment to the hasNext() method.